### PR TITLE
[SYCL] Implement multiple tracing levels for SYCL_UR_TRACE

### LINKED ISSUE #14926: implement UR_LOG_TRACING env var in UR instead of sycl RT, currently it's implemented in ur.hpp - first raised in this comment

### PR DESCRIPTION
**Pull request by callumfare**
This undoes some changes to tracing that were made when porting from PI to UR, and means that `SYCL_UR_TRACE` now has the same behavior as the `SYCL_PI_TRACE` option that it replaced.

Also only enable the UR tracing layer when `SYCL_UR_TRACE` is set to the appropriate level, rather than always enabling it when `UR_LOG_TRACING` is set - fixes #14926

### CODE REVIEW

File: sycl/source/detail/global_handler.cpp [hunk 1]
```diff
@@ -362,7 +362,9 @@ extern "C" __SYCL_EXPORT BOOL WINAPI DllMain(HINSTANCE hinstDLL,
     std::exit(1);
   }
 
-  bool PrintUrTrace = sycl::detail::ur::trace();
+  bool PrintUrTrace =
+    sycl::detail::ur::trace(sycl::detail::ur::TraceLevel::TRACE_CALLS;
```
Comment (aelovikov-intel) [inline sycl/source/detail/global_handler.cpp:366]: Where is the matching `)`?

**Comment by omarahmed1111** [comment]
@intel/dpcpp-doc-reviewers  @intel/dpcpp-esimd-reviewers @intel/dpcpp-tools-reviewers @intel/sycl-graphs-reviewers please review, Thanks!

File: sycl/doc/EnvironmentVariables.md [hunk 2]
```diff
@@ -231,6 +231,18 @@ variables in production code.</span>
...
+
+### `SYCL_UR_TRACE` Options
+
+`SYCL_UR_TRACE` accepts a bit-mask. Supported tracing levels are in the table below
+
+| Option | Description |
+| ------ | ----------- |
+| 1 | Enable basic tracing, which is tracing of UR adapters/devices discovery |
+| 2 | Enable tracing of the UR calls |
+| -1 | Enable all levels of tracing |
```
Comment (sarnex) [inline sycl/doc/EnvironmentVariables.md:243]: I don't understand the difference between `2` and `-1`

**Review by sarnex — APPROVED** [review]
lgtm minus a question

File: sycl/doc/EnvironmentVariables.md (hunk 2 above)
Comment (callumfare) [reply sycl/doc/EnvironmentVariables.md:243]: `2` enables just the tracing of UR calls, but `-1` enables tracing of the UR calls as well as the basic device/adapter discovery tracing, i.e. `1` and `2` combined (and implicitly also any future tracing levels that may be added).

Comment (sarnex) [reply sycl/doc/EnvironmentVariables.md:243]: think we can add that explanation to the doc?

**Review by EwanC — APPROVED** [review]

**Review by sarnex — APPROVED** [review]

**Comment by omarahmed1111** [comment]
@intel/dpcpp-doc-reviewers Please review, Thanks!

**Review by AlexeySachkov — APPROVED** [review]

**Review by omarahmed1111 — APPROVED** [review]

**Comment by omarahmed1111** [comment]
@intel/llvm-reviewers-runtime Please review, Thanks!

**Comment by aelovikov-intel** [comment]
To be honest, it doesn't make sense to me to have any UR traces inside SYCL RT. Is that bad naming or some bigger issue with tracing design?

**Comment by dm-vodopyanov** [comment]
@callumfare please follow https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#commit-message

**Comment by kbenzie** [comment]
> To be honest, it doesn't make sense to me to have any UR traces inside SYCL RT. Is that bad naming or some bigger issue with tracing design?

This is restoring the behavior of `SYCL_PI_TRACE`. It arguably shouldn't have been used in all the places it was either.

**Comment by callumfare** [comment]
> To be honest, it doesn't make sense to me to have any UR traces inside SYCL RT. Is that bad naming or some bigger issue with tracing design?

@aelovikov-intel The tracing is implemented in UR, but the environment variables to enable it are quite verbose:
`UR_ENABLE_LAYERS=UR_LAYER_TRACING UR_LOG_TRACING=level:info;output:stdout;flush:info UR_LOG_LOADER=level:info;output:stdout;flush:info`

The `SYCL_UR_TRACE` option is just a shortcut for this, and also provides some continuity from `SYCL_PI_TRACE` which was recently removed

**Comment by aelovikov-intel** [comment]
> This is restoring the behavior of `SYCL_PI_TRACE`. It arguably shouldn't have been used in all the places it was either.

But previously plugins were part of this project, so while not ideal implementation-wise, still not leaking abstractions terribly. I don't see that being true now that UR is separate.

**Comment by aelovikov-intel** [comment]
> The `SYCL_UR_TRACE` option is just a shortcut for this, and also provides some continuity from `SYCL_PI_TRACE` which was recently removed

Why can't that shortcut be provided at the UR side?

**Comment by callumfare** [comment]
> Why can't that shortcut be provided at the UR side?

@aelovikov-intel It could, although from a SYCL user's perspective it makes sense for the option to start with the `SYCL_` prefix, and having an option with that name in UR wouldn't be ideal. Also as Benie said above, some elements of the tracing are implemented in SYCL-RT, which does mean that the option has a misleading name, so something should probably change.

I'd rather get this merged asap to get rid of functional differences with SYCL_PI_TRACE, and then we could revisit the design of the tracing in future.

**Review by aelovikov-intel — APPROVED** [review]
Not a fan of it, but if it's needed to unblock folks...

**Pull request merged**